File: Try_GameEngine5/src/com/example/try_gameengine/remotecontroller/RemoteController.java
package com.example.try_gameengine.remotecontroller;

import java.util.ArrayList;
import java.util.List;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.view.MotionEvent;
import com.example.try_gameengine.utils.GameTimeUtil;

/**
 * 
 * @author irons
 *
 */
public class RemoteController implements IRemoteController{
	
	private static RemoteController remoteController;
	private static RemoteControl remoteControl;
	private static RemoteLoader remoteLoader;
	
	RemoteContollerListener remoteContollerListener;
	RemoteContollerOnTouchEventListener remoteContollerOnTouchEventListener;
	List<CommandType> commandTypes = new ArrayList<CommandType>();
	GameTimeUtil remoteControllerTimeUtil;
	
	public enum CommandType{
		None,
		UPKeyUpCommand,
		UPKeyDownCommand,
		DownKeyUpCommand,
		DownKeyDownCommand,
		LeftKeyUpCommand,
		LeftKeyDownCommand,
		RightKeyUpCommand,
		RightKeyDownCommand
	}
	
	public interface RemoteContollerOnTouchEventListener{
		
		public boolean onTouchEvent(MotionEvent event);
				
	}
	
	private static final int INVALID_POINTER_ID = -1;

	// The active pointer is the one currently moving our object.
	private int mActivePointerId = INVALID_POINTER_ID;
	
	private RemoteContollerOnTouchEventListener defaultRemoteContollerOnTouchEventListener = new RemoteContollerOnTouchEventListener() {
		
		@Override
		public boolean onTouchEvent(MotionEvent event) {
			// TODO Auto-generated method stub
			boolean isCatchTouchEvent = false;
			float x = event.getX();
			float y = event.getY();
			int action = event.getAction();
			switch (action & MotionEvent.ACTION_MASK) {
		    	case MotionEvent.ACTION_DOWN: 
		    		mActivePointerId = event.getPointerId(0);
		    		isCatchTouchEvent = pressDown(x, y, mActivePointerId, event);
		    		break;
		    	case MotionEvent.ACTION_POINTER_DOWN: 
		    		final int downPointerIndex = (action & MotionEvent.ACTION_POINTER_INDEX_MASK) 
                    >> MotionEvent.ACTION_POINTER_INDEX_SHIFT;
                    mActivePointerId = event.getPointerId(downPointerIndex);
                    x = event.getX(downPointerIndex);
	                y = event.getY(downPointerIndex);
	                isCatchTouchEvent = pressDown(x, y, mActivePointerId, event);
		    		break;
		    	case MotionEvent.ACTION_UP: 
		    		mActivePointerId = event.getPointerId(0);
		    		isCatchTouchEvent = pressUp(x, y, mActivePointerId, event);
		            break;
		            
		        case MotionEvent.ACTION_CANCEL: 
		            mActivePointerId = INVALID_POINTER_ID;
		            break;
		        
		        case MotionEvent.ACTION_POINTER_UP: 
		            // Extract the index of the pointer that left the touch sensor
		            final int pointerIndex = (action & MotionEvent.ACTION_POINTER_INDEX_MASK) 
		                    >> MotionEvent.ACTION_POINTER_INDEX_SHIFT;
		            final int pointerId = event.getPointerId(pointerIndex);
		            x = event.getX(pointerIndex);
	                y = event.getY(pointerIndex);
	                isCatchTouchEvent = pressUp(x, y, pointerId, event);
		            break;
			}
			return isCatchTouchEvent;
		}
	};
	
	public interface RemoteContollerListener{
		
		public void pressDown(List<CommandType> commandTypes);
				
	}
	
	private RemoteContollerListener defaultRemoteContollerListener = new RemoteContollerListener() {
		
		@Override
		public void pressDown(List<CommandType> commandTypes) {
			// TODO Auto-generated method stub
			for(CommandType commandType : commandTypes){
				switch (commandType) {
				case UPKeyUpCommand:
					
					break;
				case UPKeyDownCommand:
					
					break;
				case DownKeyUpCommand:
									
					break;
				case DownKeyDownCommand:
					
					break;
				case LeftKeyUpCommand:
					
					break;
				case LeftKeyDownCommand:
					
					break;
				case RightKeyUpCommand:
					/*//Demo. press DownKey and move Layers which are in the LayerManager go right.
					for(List<ILayer> layers : LayerManager.getLayerLevelList()){
						for(ILayer layer : layers){
							if(layer instanceof Sprite){
								((Sprite)layer).move(10,0);
							}
						}
					}*/
					break;
				case RightKeyDownCommand:
					
					break;
				default:
					break;
				}
			}			
		}
		
	};

	private RemoteController(){
		remoteLoader = new RemoteLoader();
		remoteControl = remoteLoader.getRemoteControl();
		remoteControllerTimeUtil = new GameTimeUtil(0);
		remoteContollerListener = defaultRemoteContollerListener;
		remoteContollerOnTouchEventListener = defaultRemoteContollerOnTouchEventListener;
	};
	
	public static RemoteController createRemoteController(){
		if(remoteController==null){
			synchronized (RemoteController.class) {
				if(remoteController==null)
					remoteController = new RemoteController();
			}			
		}
		return remoteController;
	}
	
	public boolean pressDown(float x, float y, int motionEventPointerId, MotionEvent event){
		CommandType commandType = remoteControl.executePressDown(x, y, motionEventPointerId, event);
		commandTypes.add(commandType);
		
		if(remoteControllerTimeUtil.isArriveExecuteTime()){
			remoteContollerListener.pressDown(commandTypes);
			commandTypes.clear();
		}
		
		if(commandType==CommandType.None){
			return false;
		}else{
			return true;
		}
	}
	
	public boolean pressUp(float x, float y, int motionEventPointerId, MotionEvent event){
		CommandType commandType = remoteControl.executePressUp(x, y, motionEventPointerId, event);
		commandTypes.add(commandType);
		
		if(remoteControllerTimeUtil.isArriveExecuteTime()){
			remoteContollerListener.pressDown(commandTypes);
			commandTypes.clear();
		}
		
		if(commandType==CommandType.None){
			return false;
		}else{
			return true;
		}
	}
	
	public void execute(){
		this.remoteContollerListener.pressDown(commandTypes);
	}
	
	public void setRemoteContollerListener(RemoteContollerListener remoteContollerListener){
		this.remoteContollerListener = remoteContollerListener;
	}
	
	public void setRemoteContollerOnTouchEventListener(RemoteContollerOnTouchEventListener remoteContollerOnTouchEventListener){
		this.remoteContollerOnTouchEventListener = remoteContollerOnTouchEventListener;
	}
	
	public boolean onTouchEvent(MotionEvent event){
		return this.remoteContollerOnTouchEventListener.onTouchEvent(event);
	}
	
	public UpKey getUpKey(){
		return remoteLoader.getUpKey();
	}
	
	public DownKey getDownKey(){
		return remoteLoader.getDownKey();
	}
	
	public LeftKey getLeftKey(){
		return remoteLoader.getLeftKey();
	}
	
	public RightKey getRightKey(){
		return remoteLoader.getRightKey();
	}
	
	public void drawRemoteController(Canvas canvas, Paint paint){
//		remoteLoader.getUpKey().drawSelf(canvas, paint);
//		remoteLoader.getDownKey().drawSelf(canvas, paint);
		remoteLoader.getLeftKey().drawSelf(canvas, paint);
		remoteLoader.getRightKey().drawSelf(canvas, paint);
	}
	
	public void setUpKyPosition(float x, float y){
		remoteLoader.getUpKey().setPosition(x, y);
	}
	
	public void setUpKyBitmap(Bitmap bitmap){
		remoteLoader.getUpKey().setBitmapAndAutoChangeWH(bitmap);
	}
	
	public void setDownKyPosition(float x, float y){
		remoteLoader.getDownKey().setPosition(x, y);
	}
	
	public void setDownKyBitmap(Bitmap bitmap){
		remoteLoader.getDownKey().setBitmapAndAutoChangeWH(bitmap);
	}
	
	public void setLeftKyPosition(float x, float y){
		remoteLoader.getLeftKey().setPosition(x, y);
	}
	
	public void setLeftKyBitmap(Bitmap bitmap){
		remoteLoader.getLeftKey().setBitmapAndAutoChangeWH(bitmap);
	}
	
	public void setRightKyPosition(float x, float y){
		remoteLoader.getRightKey().setPosition(x, y);
	}
	
	public void setRightKyBitmap(Bitmap bitmap){
		remoteLoader.getRightKey().setBitmapAndAutoChangeWH(bitmap);
	}
}

File: Try_GameEngine5/src/com/example/try_gameengine/remotecontroller/RemoteController.kt
package com.example.try_gameengine.remotecontroller

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.view.MotionEvent
import com.example.try_gameengine.utils.GameTimeUtil

/**
 * 
 * @author irons
 // */
class RemoteController private constructor() : IRemoteController {
    @JvmField
    var remoteContollerListener: RemoteContollerListener
    @JvmField
    var remoteContollerOnTouchEventListener: RemoteContollerOnTouchEventListener
    var commandTypes: MutableList<CommandType?> = ArrayList<CommandType?>()
    var remoteControllerTimeUtil: GameTimeUtil

    enum class CommandType {
        None,
        UPKeyUpCommand,
        UPKeyDownCommand,
        DownKeyUpCommand,
        DownKeyDownCommand,
        LeftKeyUpCommand,
        LeftKeyDownCommand,
        RightKeyUpCommand,
        RightKeyDownCommand
    }

    interface RemoteContollerOnTouchEventListener {
        fun onTouchEvent(event: MotionEvent?): Boolean
    }

    // The active pointer is the one currently moving our object.
    private var mActivePointerId: Int = INVALID_POINTER_ID

    private val defaultRemoteContollerOnTouchEventListener: RemoteContollerOnTouchEventListener =
        object : RemoteContollerOnTouchEventListener {
            override fun onTouchEvent(event: MotionEvent?): Boolean {
                event ?: return false
                // TODO Auto-generated method stub
                var isCatchTouchEvent = false
                var x = event.getX()
                var y = event.getY()
                val action = event.getAction()
                when (action and MotionEvent.ACTION_MASK) {
                    MotionEvent.ACTION_DOWN -> {
                        mActivePointerId = event.getPointerId(0)
                        isCatchTouchEvent = pressDown(x, y, mActivePointerId, event)
                    }

                    MotionEvent.ACTION_POINTER_DOWN -> {
                        val downPointerIndex = ((action and MotionEvent.ACTION_POINTER_INDEX_MASK)
                                shr MotionEvent.ACTION_POINTER_INDEX_SHIFT)
                        mActivePointerId = event.getPointerId(downPointerIndex)
                        x = event.getX(downPointerIndex)
                        y = event.getY(downPointerIndex)
                        isCatchTouchEvent = pressDown(x, y, mActivePointerId, event)
                    }

                    MotionEvent.ACTION_UP -> {
                        mActivePointerId = event.getPointerId(0)
                        isCatchTouchEvent = pressUp(x, y, mActivePointerId, event)
                    }

                    MotionEvent.ACTION_CANCEL -> mActivePointerId = INVALID_POINTER_ID
                    MotionEvent.ACTION_POINTER_UP -> {
                        // Extract the index of the pointer that left the touch sensor
                        val pointerIndex = ((action and MotionEvent.ACTION_POINTER_INDEX_MASK)
                                shr MotionEvent.ACTION_POINTER_INDEX_SHIFT)
                        val pointerId = event.getPointerId(pointerIndex)
                        x = event.getX(pointerIndex)
                        y = event.getY(pointerIndex)
                        isCatchTouchEvent = pressUp(x, y, pointerId, event)
                    }
                }
                return isCatchTouchEvent
            }
        }

    interface RemoteContollerListener {
        fun pressDown(commandTypes: MutableList<CommandType?>?)
    }

    private val defaultRemoteContollerListener: RemoteContollerListener =
        object : RemoteContollerListener {
            override fun pressDown(commandTypes: MutableList<CommandType?>?) {
                // TODO Auto-generated method stub
                commandTypes ?: return
                for (commandType in commandTypes) {
                    when (commandType) {
                        CommandType.UPKeyUpCommand -> {}
                        CommandType.UPKeyDownCommand -> {}
                        CommandType.DownKeyUpCommand -> {}
                        CommandType.DownKeyDownCommand -> {}
                        CommandType.LeftKeyUpCommand -> {}
                        CommandType.LeftKeyDownCommand -> {}
                        CommandType.RightKeyUpCommand -> {}
                        CommandType.RightKeyDownCommand -> {}
                        else -> {}
                    }
                }
            }
        }

    init {
        remoteLoader = RemoteLoader()
        remoteControl = remoteLoader.getRemoteControl()
        remoteControllerTimeUtil = GameTimeUtil(0)
        remoteContollerListener = defaultRemoteContollerListener
        remoteContollerOnTouchEventListener = defaultRemoteContollerOnTouchEventListener
    }

    fun pressDown(x: Float, y: Float, motionEventPointerId: Int, event: MotionEvent?): Boolean {
        val commandType: CommandType? =
            remoteControl.executePressDown(x, y, motionEventPointerId, event)
        commandTypes.add(commandType)

        if (remoteControllerTimeUtil.isArriveExecuteTime) {
            remoteContollerListener.pressDown(commandTypes)
            commandTypes.clear()
        }

        if (commandType == CommandType.None) {
            return false
        } else {
            return true
        }
    }

    fun pressUp(x: Float, y: Float, motionEventPointerId: Int, event: MotionEvent?): Boolean {
        val commandType: CommandType? =
            remoteControl.executePressUp(x, y, motionEventPointerId, event)
        commandTypes.add(commandType)

        if (remoteControllerTimeUtil.isArriveExecuteTime) {
            remoteContollerListener.pressDown(commandTypes)
            commandTypes.clear()
        }

        if (commandType == CommandType.None) {
            return false
        } else {
            return true
        }
    }

    fun execute() {
        this.remoteContollerListener.pressDown(commandTypes)
    }

    fun setRemoteContollerListener(remoteContollerListener: RemoteContollerListener) {
        this.remoteContollerListener = remoteContollerListener
    }

    fun setRemoteContollerOnTouchEventListener(remoteContollerOnTouchEventListener: RemoteContollerOnTouchEventListener) {
        this.remoteContollerOnTouchEventListener = remoteContollerOnTouchEventListener
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        return this.remoteContollerOnTouchEventListener.onTouchEvent(event)
    }

    fun getUpKey(): UpKey? {
        return remoteLoader.getUpKey()
    }

    fun getDownKey(): DownKey? {
        return remoteLoader.getDownKey()
    }

    fun getLeftKey(): LeftKey? {
        return remoteLoader.getLeftKey()
    }

    fun getRightKey(): RightKey? {
        return remoteLoader.getRightKey()
    }

    override fun drawRemoteController(canvas: Canvas?, paint: Paint?) {
//		remoteLoader.getUpKey().drawSelf(canvas, paint);
//		remoteLoader.getDownKey().drawSelf(canvas, paint);
        remoteLoader.getLeftKey()!!.drawSelf(canvas, paint)
        remoteLoader.getRightKey()!!.drawSelf(canvas, paint)
    }

    fun setUpKyPosition(x: Float, y: Float) {
        remoteLoader.getUpKey()!!.setPosition(x, y)
    }

    fun setUpKyBitmap(bitmap: Bitmap) {
        remoteLoader.getUpKey()!!.setBitmapAndAutoChangeWH(bitmap)
    }

    fun setDownKyPosition(x: Float, y: Float) {
        remoteLoader.getDownKey()!!.setPosition(x, y)
    }

    fun setDownKyBitmap(bitmap: Bitmap) {
        remoteLoader.getDownKey()!!.setBitmapAndAutoChangeWH(bitmap)
    }

    fun setLeftKyPosition(x: Float, y: Float) {
        remoteLoader.getLeftKey()!!.setPosition(x, y)
    }

    fun setLeftKyBitmap(bitmap: Bitmap) {
        remoteLoader.getLeftKey()!!.setBitmapAndAutoChangeWH(bitmap)
    }

    fun setRightKyPosition(x: Float, y: Float) {
        remoteLoader.getRightKey()!!.setPosition(x, y)
    }

    fun setRightKyBitmap(bitmap: Bitmap) {
        remoteLoader.getRightKey()!!.setBitmapAndAutoChangeWH(bitmap)
    }

    companion object {
        private var remoteController: RemoteController? = null
        private lateinit var remoteControl: RemoteControl
        private lateinit var remoteLoader: RemoteLoader

        private val INVALID_POINTER_ID = -1

        fun createRemoteController(): RemoteController {
            if (remoteController == null) {
                synchronized(RemoteController::class.java) {
                    if (remoteController == null) remoteController = RemoteController()
                }
            }
            return remoteController!!
        }
    }
}
